pkg/gpu/context.go

```diff
@@ -209,13 +218,67 @@ func (ctx *systemContext) cleanupOldEntries() {
 	}
 }
 
+// filterDevicesForContainer filters the available GPU devices for the given
+// container. If the ID is not empty, we check the assignment of GPU resources
+// to the container and return only the devices that are available to the
+// container.
+func (ctx *systemContext) filterDevicesForContainer(devices []nvml.Device, containerID string) []nvml.Device {
+	if containerID == "" {
+		// If the process is not running in a container, we assume all devices are available.
+		return devices
+	}
+
+	container, err := ctx.workloadmeta.GetContainer(containerID)
+	if err != nil {
+		if !errors.IsNotFound(err) {
+			log.Warnf("Error getting container %s: %s", containerID, err)
+		}
+
+		return devices
```

Comment:
My initial idea was that it was better to send metrics that could sometimes be inaccurate (they won't always be inaccurate even in case of errors) rather than fail to send things completely. But I'm not sure what's the usual practice, it's true that it inaccurate metrics can be confusing but also missed metrics could be too. Maybe sending an extra metric called "problems while calculating metrics" could alert the user of these cases?